### PR TITLE
Add support for NXP SGTL5000, Fe-Pi Audio Sound Card 

### PR DESCRIPTION
**Pull request by Fe-Pi**
Fe-Pi Audio Sound Card is based on NXP(Freescale) SGTL5000 codec.
Mechanical specification of the board is the same the Raspberry Pi Zero.
3.5mm jacks for Headphone/Mic, Line In, and Line Out.

For your reference, the following link takes you to pictures of the working prototype board.
https://goo.gl/photos/BdPfK8vjyQDPtSmCA

### CODE REVIEW

File: sound/soc/bcm/fe-pi-audio.c [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	
+	of_node_put(i2s_node);
+	
+	card->dev = &pdev->dev;
+	platform_set_drvdata(pdev, card);
+	
+	ret = snd_soc_register_card(card);
+	if (ret == -EPROBE_DEFER) return ret;
+	if (ret) dev_err(&pdev->dev, "snd_soc_register_card() failed: %d\n", ret);
+	if (!ret) dev_info(&pdev->dev, "snd_soc_register_card() OK\n");
```
Comment (clivem) [inline sound/soc/bcm/fe-pi-audio.c]: Maybe make the dev_info logging a dev_dbg.....

**Comment by Fe-Pi** [comment]
changed dev_info to dev_dbg

**Comment by clivem** [comment]
> changed dev_info to dev_dbg

Looks like you changed it to dev_err, not dev_dbg.

**Comment by Fe-Pi** [comment]
My mistake, changed to dev_dbg.

File: arch/arm/boot/dts/overlays/README [hunk 2]
```diff
@@ -305,6 +305,10 @@ Params: int_pin                 GPIO used for INT (default 25)
 
         speed                   SPI bus speed (default 12000000)
 
+Name:   fe-pi-audio
+Info:   Configures the Fe-Pi Audio Sound Card
+Load:   dtoverlay=fe-pi-audio
+Params: <None>
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README:319]: Extra blank line needed here.

File: arch/arm/boot/dts/overlays/README [hunk 3]
```diff
@@ -305,6 +305,10 @@ Params: int_pin                 GPIO used for INT (default 25)
 
         speed                   SPI bus speed (default 12000000)
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README:313]: Extra blank line needed here.

File: sound/soc/bcm/Kconfig [hunk 4]
```diff
@@ -85,3 +85,12 @@ config SND_BCM2708_SOC_ADAU1977_ADC
 	select SND_SOC_ADAU1977_I2C
 	help
 	  Say Y or M if you want to add support for ADAU1977 ADC.
+
+config SND_BCM2708_SOC_FE_PI_AUDIO
+	tristate "Support for Fe-Pi-Audio"
+	depends on SND_BCM2708_SOC_I2S || SND_BCM2835_SOC_I2S
+	select SND_SOC_SGTL5000
+	select REGULATOR
```
Comment (pelwell) [inline sound/soc/bcm/Kconfig:100]: We don't use regulators in the Pi kernel; try not defining these CONFIG setting - you may find that it just works.

**Comment by pelwell** [comment]
I've added a few inline comments. You'll also need to rebase against the latest commits.
If you can address these issues, squash your commits down to one and force push (git push -f) then I'll merge it.

File: sound/soc/bcm/Kconfig (hunk 4 above)
Comment (Fe-Pi) [inline sound/soc/bcm/Kconfig:100]: The SGTL5000 codec driver needs regulator. I tried without it, did not work. Got the following without regulator....

[   14.331875] sgtl5000 1-000a: sgtl5000 revision 0x11
[   14.414956] sgtl5000 1-000a: this setup needs regulator support in the kernel
[   14.414991] sgtl5000 1-000a: Failed to register vddd internal supplies: -22
[   14.415004] sgtl5000 1-000a: ASoC: failed to probe component -22
[   14.415061] snd-fe-pi-audio soc:sound: ASoC: failed to instantiate card -22
[   14.415156] snd-fe-pi-audio soc:sound: snd_soc_register_card() failed: -22
[   14.415193] snd-fe-pi-audio: probe of soc:sound failed with error -22

**Comment by Fe-Pi** [comment]
Done, Thank you!

**Comment by pelwell** [comment]
@popcornmix I know we don't use regulators on the Pi, but the SGTL5000 driver annoyingly uses the regulator framework to control it's own regulators.

Enabling CONFIG_REGULATOR seems to eat up 188KiB in total (figured obtained by running top for several minutes on a quiescent system with both configurations and taking the maximum free memory in both cases).

I think there are four options:
1) allow this patch, and enabled CONFIG_REGULATOR,
2) require the SGTL5000 driver to be modified so as not to need regulator framework,
3) require a driver-specific library that emulates a minimal regulator framework (effectively a variant of 2),
4) refuse this patch.

What are your thoughts?

**Comment by popcornmix** [comment]
CONFIG_REGULATOR may cause issues. It's presence broke HifiBerry support on OpenELEC some time back:
http://forum.kodi.tv/showthread.php?tid=192380&pid=1836596#pid1836596
Not sure if that is still the case, but it will certainly open up new code paths in some existing drivers, so will need some care.

**Comment by clivem** [comment]
@popcornmix I have a vague recollection, that if you enable CONFIG_REGULATOR you need to add a couple of FIXED regs to dt for DVDD/PVDD @ 3.3V, for wm8804 driver, for HB digi. Somewhere, I have a patch for hifiberry_digi overlay. I used to build all my kernels with CONFIG_REGULATOR enabled.

**Comment by Fe-Pi** [comment]
@clivem @pelwell @popcornmix Any ideas on what I can do to help move this PR forward, if possible, Thanks.

**Pull request closed without merging**